api/client/src/main/java/org/projectnessie/client/auth/oauth2/AbstractFlow.java

```diff
@@ -27,15 +27,31 @@ abstract class AbstractFlow implements Flow {
     this.config = config;
   }
 
-  <T> T invokeEndpoint(URI endpoint, Object request, Class<? extends T> responseClass) {
+  <REQ extends TokensRequestBase, RESP extends TokensResponseBase> RESP invokeTokenEndpoint(
+      TokensRequestBase.Builder<REQ> request, Class<? extends RESP> responseClass) {
+    request.scope(config.getScope().orElse(null));
+    if (request instanceof PublicTokensRequestBase.Builder
+        && !config.getClientSecret().isPresent()) {
+      ((PublicTokensRequestBase.Builder<?>) request).clientId(config.getClientId());
+    }
+    return invokeEndpoint(config.getResolvedTokenEndpoint(), request.build(), responseClass);
+  }
+
+  DeviceCodeResponse invokeDeviceAuthEndpoint() {
+    DeviceCodeRequest.Builder request =
+        DeviceCodeRequest.builder().scope(config.getScope().orElse(null));
+    if (!config.getClientSecret().isPresent()) {
```

Comment:
Same question here.

Reply:
Basically: include the client ID for public clients. Let me refactor this a bit to clarify.